### ISSUE TITLE
[Change Request] array-bracket-spacing: "always" behaviour

### ISSUE THREAD

**Issue by DanielSmedegaardBuus** [eslint/eslint#8922]
Hi :) First of all, this is not a bug report, it's a behavioural change request, but the link given to request changes directs back here, so...

**Tell us about your environment**
SublimeText 3 on MacOS

* **ESLint Version:**
v3.17.1

* **Node Version:**
v8.1.2

* **npm Version:**
5.0.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
<details>

```json
{
  "space": 2,
  "env": {
    "es6": true,
    "node": true,
    "mocha": true
  },
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true,
      "experimentalObjectRestSpread": true,
      "impliedStrict": true
    }
  },
  "ignores": [
    "html/*"
  ],
  "extends": "eslint:recommended",
  "rules": {
    "linebreak-style": [ "error", "unix" ],
    "space-before-function-paren": "error",
    "no-console": "off",
    "semi": [ "error", "always" ],
    "object-curly-spacing": [ "warn", "always" ],
    "key-spacing": [
      "error", {
        "align": {
          "beforeColon": true,
          "afterColon": true,
          "on": "colon"
        },
        "singleLine": {
          "beforeColon": false,
          "afterColon": true
        }
      }
    ],
    "array-bracket-spacing": [ "warn", "always",
      { "arraysInArrays": false, "objectsInArrays": false }
    ],
    "eol-last": [ "off" ],
    "unicorn/filename-case": [ "off" ],
    "curly": [ "error", "multi-or-nest" ]
  }
}
```
<summary>Configuration</summary>

```js



```

</details>

**What did you do? Please include the actual source code causing the issue.**
I've used the `"objectsInArrays": false` modifier to allow having no spaces when you're doing stuff like

```javascript
var a = [{
  propsy : 1,
  horses : "big",
  a      : "b"
}, {
  propsy : 2,
  ponies : "not quite as big but still big",
  a      : "c"
}];
```

This is often useful to save an indentation level to avoid linebreaks when trying to stay within an 80-character width.

However, this rule then warns when I do stuff like

```javascript
var a = [ 1, 2, 'horses', {} ]; // Last space disallowed
var a = [ 1, 2, 'horses', {}];  // No warnings, but looks weird :D
```

```js





```

**What did you expect to happen?**
I sort of expected that `"objectsInArrays": false` would just make the `array-bracket-spacing` rule not be enforced with regards to objects, instead of making it enforce the opposite rule... But seeing as how that would then not warn when doing `[ 1, 2, {}]` I guess it's not feasible to do the way I'd like it to... I don't know, maybe it's not an issue. 

**What actually happened? Please include the actual, raw output from ESLint.**

**Comment by not-an-aardvark** [comment]
Thanks for reporting. I'm not quite sure what you're proposing -- could you clarify?

**Comment by platinumazure** [comment]
@DanielSmedegaardBuus Are you suggesting that `objectsInArrays: false` should only enforce lack of space if the array just has

* One object?
    ```js
    const foo = [{ a: true }];
    ```
* Only objects?
    ```js
    const foo = [{}, {}];
    ```
* Only if first and last item are objects?
    ```js
    const foo = [{}, true, {}];
    ```

Or are you asking about something completely different?

**Comment by DanielSmedegaardBuus** [comment]
Hi guys :) I guess It's the either-or rule that's confusing me.

Like, here I'm telling eslint that I'd like arrays to have a space after the opening and before the closing bracket, but then realising that it looks weird when I start or end an array with an object or another array.

So I configure the `objectsInArrays: false` setting, thinking that that allows me to omit the space in case of objects inside the array, not that I'm now **forcing** the omission of spaces for that scenario.

I'm not entirely sure what I'm requesting then, really... Sorry :/ This seems to be how the rules work in general, so I'm not gonna request a change to that :) But if anything reasonable to request, then yes, some rule to allow outer brackets/curlies to "stack together" as in all of the the examples from @platinumazure would be nice. Something like `allowConsecutiveCurliesAndBrackets: true` would be awesome :)

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
